### DISCUSSION TITLE
Info: Nous A1T Tasmota 2023 from Factory Dis-Assembly

### DISCUSSION THREAD

**Discussion by savejeff** [arendst/Tasmota#20479] [General]
Hi, I just bought a 4-pack of Nous A1T with pre-installed Tasmota with ESP8266.
I was planning to write my own firmware for a side project and found some older pictures [here](https://templates.blakadder.com/nous_A1.html) where it looked like the A1T are easily dis-assembled and had easy to access pads for Programming. 

sadly after opening it up, the situation I found is the opposite. Thus I wanted to put the information out there
1.) There is a hidden screw with a non-standard head
2.) The antenna is not detachable and is glued in
3.) The PCB and electronics are press-fitted into insert-style hole on the bottom. once pressed out, they get looser
4.) And worst of all: there are no programming/debugging pins on the PCB. so when firmware upload fails, it's bricked. no serial programming possible

![PXL_20240112_172856951](https://github.com/arendst/Tasmota/assets/11009455/6b571036-0340-4317-8b5c-7f85c47327de)

![PXL_20240112_165857288](https://github.com/arendst/Tasmota/assets/11009455/a3aed6a6-fb45-476d-83c3-716a8d5a8664)

![PXL_20240112_165942774](https://github.com/arendst/Tasmota/assets/11009455/2afef2fe-a5cf-42df-8f25-5753c87dbe07)

![PXL_20240112_165931894](https://github.com/arendst/Tasmota/assets/11009455/433010fe-f0c0-4aff-a80b-ad0739762e06)

![PXL_20240112_165908068](https://github.com/arendst/Tasmota/assets/11009455/5252eb17-be2c-4c7f-8ce5-271085fca16e)

![PXL_20240112_170043413](https://github.com/arendst/Tasmota/assets/11009455/78acfa1a-8f63-4f18-a821-16ce4bfb5969)

sorry for the bad pictures. i already put it all together after noticing how bad they are and I'm not taking it apart again

**Comment by s-hadinger** [comment]
Thanks for the detailed information. Plan B could be to extend Tasmota with your own features and benefit from the solid OTA features of Tasmota ;)

**Reply by savejeff** [reply]
No Problem. I mainly wanted to get the information out there for others, as it was significantly different from what i was expecting.

Ill develop the firmware on an older Gosund EP2 with accessible programming pins. Developing firmware without serial output is to error prone from me

**Comment by sfromis** [comment]
https://github.com/tasmota/ESP8266MinimalHTTPUpdater/ could be used to OTA to something non-Tasmota once, but then your project would need solid OTA support if you'd like to be able to update it.

**Reply by savejeff** [reply]
Yes thats basicly my plan right now. I have an older Gosund EP2 that has the same pin configuration but the programming pins are accessible. Ill develop my firmware there with working OTA. only then ill start flashing the Nous A1T

One Question: Is there any restrictions to what i can upload in the firmware update section of the webserver running on Tasmota devices? I've seen the tasmota.bin.bz are just the compressed firmware.bin. So if i match the flash size it should just switch over to my firmware right?

**Reply by sfromis** [reply]
Basically, IDK. However, as I do observe that the library used to do the OTA upgrading is part of the Arduino project, my guess would be that "normal" Arduino conventions apply, including the handling of `.gz` binaries.
https://github.com/esp8266/Arduino/tree/master/libraries/ESP8266HTTPUpdateServer

Browsing the minimal HTTP updater project, you can see how simple it really is.

**Reply by savejeff** [reply]
Im quite familiar with the Update functionality of ESP32 and ESP8266. 
my question was rather, if tasmota checks an incoming firmware file before accepting it. also what happens if partitions do not match between the installed firmware and the firmware to be updated to. I've seen the Tasmota project has its own board and partition definitions.

I guess ill test this with the EP2 that i can reprogram. ill write a current Tasmota firmware on it and try to upload my own firmware file.

**Reply by Noschvie** [reply]
What are you missing in Tasmota that you need "your Firmware"?

**Reply by savejeff** [reply]
Im planing a couple of things (still evolving)
- higher send rate in the range of 1s to 2.5s between publish
- flexible publish rates with higher rates when actually in use and slower in idle
- turn off the led after some time as i find it annoying to have it light up or blink all night
- have functionality like hold or double click to only turn on the relay for a set period and the turn off again
- i would like to explore UDP and some basic AES encryption to remove overhead from MQTT
- use sleep modes to save on energy and use the power sensing pulse to wake up when actual power is drained

most of that might be already possible with Tasmota, but as I said this is just a side project to learn something. I already have the firmware running with most of the above-mentioned features. I only need the OTA part

**Comment by Jason2866** [comment]
The device is not bricked when firmware is not starting. Firmware can directly programmed to the flash chip with a flash programmer. For example with a cheap ch341 flash chip programmer.

**Reply by savejeff** [reply]
okay thats good to know. never had to do this and ill try to avoid it but still good to know that there is the option.

I'm usually am using the ESP32 Modules like the WROOM and WROVER. with this flash chip programming is not possible and i never needed to do it

**Reply by Jason2866** [reply]
Wroom and Wrover modules do have most of the time standard flash chips under the metal can.
Replaced a flash chip of a Wrover module already ;-)

**Reply by savejeff** [reply]
Haha okay nice job, but at this point i would just replace the whole wrover module. if you have no components on the other side of the PCB, some hotair makes it super easy.

Okay maybe a to personal question here so just ignore if you want but I've seen you name around in the ESP32 community. In what capacity are you working with/for espressif. are you just doing this as a hobby? i like the community here on github around esp products, i think it furthers the cause of arduino and open source devices. same goes for "me-no-dev"

**Reply by Jason2866** [reply]
Everything around Tasmota and esp is just a hobby. me-no-dev is an espressif employee
